### PR TITLE
*: support JSON format output in explain statement

### PR DESCRIPTION
**Pull request by fzzf678**
### What problem does this PR solve?

Support JSON format output of explain result
Issue Number: ref #39261 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/common_plans.go [hunk 1]
```diff
@@ -922,6 +1014,42 @@ func (e *Explain) prepareOperatorInfo(p Plan, taskType, id string) {
...
+		EstRows:      estRows,
+		TaskType:     taskType,
+		AccessObject: accessObject,
+		OperatorInfo: operatorInfo,
+		Idx:          idx,
+		ChildrenIdx:  childrenIdx,
+	}
+
+	if e.Analyze || e.RuntimeStatsColl != nil {
+		if strings.ToLower(e.Format) == types.ExplainFormatVerbose || strings.ToLower(e.Format) == types.ExplainFormatTrueCardCost {
```
Comment (time-and-fate) [inline planner/core/common_plans.go:1034]: Is this case possible?

File: planner/core/common_plans.go [hunk 2]
```diff
@@ -922,6 +1014,42 @@ func (e *Explain) prepareOperatorInfo(p Plan, taskType, id string) {
...
+			jsonRow.EstCost = estCost
+		}
+		if strings.ToLower(e.Format) == types.ExplainFormatTrueCardCost {
```
Comment (time-and-fate) [inline planner/core/common_plans.go:1037]: ditto.

File: planner/core/common_plans.go [hunk 3]
```diff
@@ -922,6 +1014,42 @@ func (e *Explain) prepareOperatorInfo(p Plan, taskType, id string) {
...
+			jsonRow.CostFormula = costFormula
+		}
+		actRows, analyzeInfo, memoryInfo, diskInfo := getRuntimeInfoStr(e.ctx, p, e.RuntimeStatsColl)
+		jsonRow.ActRows = actRows
+		jsonRow.ExecuteInfo = analyzeInfo
+		jsonRow.MemoryInfo = memoryInfo
+		jsonRow.DiskInfo = diskInfo
+	} else {
+		if strings.ToLower(e.Format) == types.ExplainFormatVerbose || strings.ToLower(e.Format) == types.ExplainFormatTrueCardCost {
```
Comment (time-and-fate) [inline planner/core/common_plans.go:1046]: ditto.

**Review by time-and-fate — COMMENTED** [review]
I think we also need some test case outputs to show what the complete output be like. Not just assertions.

File: planner/core/common_plans.go [hunk 4]
```diff
@@ -649,6 +649,41 @@ type SelectInto struct {
...
+	DiskInfo            string `json:"diskInfo,omitempty"`
+	TotalMemoryConsumed string `json:"totalMemoryConsumed,omitempty"`
+}
+
+// JSONSlice store JSONExplainRows
+type JSONSlice []*ExplainInfoForEncode
+
+// ToString convert json to string
+func (j *JSONSlice) ToString() (string, error) {
+	b, err := json.Marshal(*j)
```
Comment (time-and-fate) [inline planner/core/common_plans.go:675]: I think this would make the output be `HTMLEscape`d, which is probably not expected.

**Comment by fzzf678** [comment]
@time-and-fate PTAL again

**Comment by dveeden** [comment]
It looks like the JSON explain output is somehow extremely wide. This doesn't seem to be the case for MySQL 8.0. Not a huge problem, but would be nice if we can fix this.

```
sql> explain format=json select count(*) from dual;
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| JSON                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| [
    {
        "id": "StreamAgg_6",
        "estRows": "1.00",
        "taskType": "root",
        "accessObject": "",
        "operatorInfo": "funcs:count(1)->Column#1",
        "subOperators": [
            {
                "id": "TableDual_8",
                "estRows": "1.00",
                "taskType": "root",
                "accessObject": "",
                "operatorInfo": "rows:1",
                "subOperators": []
            }
        ]
    }
]
 |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.0012 sec)

sql> explain format=json select count(*) from dual\G
*************************** 1. row ***************************
JSON: [
    {
        "id": "StreamAgg_6",
        "estRows": "1.00",
        "taskType": "root",
        "accessObject": "",
        "operatorInfo": "funcs:count(1)->Column#1",
        "subOperators": [
            {
                "id": "TableDual_8",
                "estRows": "1.00",
                "taskType": "root",
                "accessObject": "",
                "operatorInfo": "rows:1",
                "subOperators": []
            }
        ]
    }
]

1 row in set (0.0011 sec)
```

```
mysql-8.0.31> explain format=json select count(*) from dual;
+--------------------------------------------------------------------------------+
| EXPLAIN                                                                        |
+--------------------------------------------------------------------------------+
| {
  "query_block": {
    "select_id": 1,
    "message": "No tables used"
  }
} |
+--------------------------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)
```

**Comment by dveeden** [comment]
The line length of the output seems to increase quickly when adding more operations:
```
$ mysql -u root -h 127.0.0.1 -P 4000 -te 'EXPLAIN FORMAT=JSON SELECT 1' | tail -1 | wc -L
457
$ mysql -u root -h 127.0.0.1 -P 4000 -te 'EXPLAIN FORMAT=JSON SELECT 1 UNION ALL SELECT 1' | tail -1 | wc -L
1367
$ mysql -u root -h 127.0.0.1 -P 4000 -te 'EXPLAIN FORMAT=JSON SELECT 1 UNION ALL SELECT 1 UNION ALL SELECT 1' | tail -1 | wc -L
1956
```

**Comment by dveeden** [comment]
I assume it is not easily possible to follow the same structure as the JSON explain in MySQL? If we could this would make it easier for IDE's and SQL Clients (DBeaver, Datagrip, MySQL Workbench, etc) to use the JSON explain output.

We could do something like `EXPLAIN FORMAT=TI_JSON ...` to allow us to later on add a fully compatible `EXPLAIN FORMAT=JSON ...`. But I don't think we should.

I assume there is no need to implement[ MySQL's `end_markers_in_json`](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_end_markers_in_json)?

```
mysql-8.0.31> EXPLAIN FORMAT=JSON SELECT 1\G
*************************** 1. row ***************************
EXPLAIN: {
  "query_block": {
    "select_id": 1,
    "message": "No tables used"
  }
}
1 row in set, 1 warning (0.00 sec)

mysql-8.0.31> set session end_markers_in_json=on;
Query OK, 0 rows affected (0.00 sec)

mysql-8.0.31> EXPLAIN FORMAT=JSON SELECT 1\G
*************************** 1. row ***************************
EXPLAIN: {
  "query_block": {
    "select_id": 1,
    "message": "No tables used"
  } /* query_block */
}
1 row in set, 1 warning (0.00 sec)
```

**Comment by fzzf678** [comment]
> It looks like the JSON explain output is somehow extremely wide. This doesn't seem to be the case for MySQL 8.0. Not a huge problem, but would be nice if we can fix this.
> 
> ```
> sql> explain format=json select count(*) from dual;
> +-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | JSON                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
> +-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | [
>     {
>         "id": "StreamAgg_6",
>         "estRows": "1.00",
>         "taskType": "root",
>         "accessObject": "",
>         "operatorInfo": "funcs:count(1)->Column#1",
>         "subOperators": [
>             {
>                 "id": "TableDual_8",
>                 "estRows": "1.00",
>                 "taskType": "root",
>                 "accessObject": "",
>                 "operatorInfo": "rows:1",
>                 "subOperators": []
>             }
>         ]
>     }
> ]
>  |
> +-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> 1 row in set (0.0012 sec)
> 
> sql> explain format=json select count(*) from dual\G
> *************************** 1. row ***************************
> JSON: [
>     {
>         "id": "StreamAgg_6",
>         "estRows": "1.00",
>         "taskType": "root",
>         "accessObject": "",
>         "operatorInfo": "funcs:count(1)->Column#1",
>         "subOperators": [
>             {
>                 "id": "TableDual_8",
>                 "estRows": "1.00",
>                 "taskType": "root",
>                 "accessObject": "",
>                 "operatorInfo": "rows:1",
>                 "subOperators": []
>             }
>         ]
>     }
> ]
> 
> 1 row in set (0.0011 sec)
> ```
> 
> ```
> mysql-8.0.31> explain format=json select count(*) from dual;
> +--------------------------------------------------------------------------------+
> | EXPLAIN                                                                        |
> +--------------------------------------------------------------------------------+
> | {
>   "query_block": {
>     "select_id": 1,
>     "message": "No tables used"
>   }
> } |
> +--------------------------------------------------------------------------------+
> 1 row in set, 1 warning (0.00 sec)
> ```

actually, the default format show the result like a sheet, we just reuse output code and the show result in one line so this would be huge when the plan is large. Similarly  `explain format = dot` also has wide row, as you found, we recommend use `\G` in the end of sql more

**Comment by fzzf678** [comment]
> I assume it is not easily possible to follow the same structure as the JSON explain in MySQL? If we could, this would make it easier for IDE's and SQL Clients (DBeaver, Datagrip, MySQL Workbench, etc) to use the JSON explain output.
> 
> We could do something like `EXPLAIN FORMAT=TI_JSON ...` to allow us to later on add a fully compatible `EXPLAIN FORMAT=JSON ...`. But I don't think we should.
> 
> I assume there is no need to implement[ MySQL's `end_markers_in_json`](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_end_markers_in_json)?
> 
> ```
> mysql-8.0.31> EXPLAIN FORMAT=JSON SELECT 1\G
> *************************** 1. row ***************************
> EXPLAIN: {
>   "query_block": {
>     "select_id": 1,
>     "message": "No tables used"
>   }
> }
> 1 row in set, 1 warning (0.00 sec)
> 
> mysql-8.0.31> set session end_markers_in_json=on;
> Query OK, 0 rows affected (0.00 sec)
> 
> mysql-8.0.31> EXPLAIN FORMAT=JSON SELECT 1\G
> *************************** 1. row ***************************
> EXPLAIN: {
>   "query_block": {
>     "select_id": 1,
>     "message": "No tables used"
>   } /* query_block */
> }
> 1 row in set, 1 warning (0.00 sec)
> ```

yeah, but this requirement is from internal, we just wrap the `format = row` result in JSON now and still can append lot of info(cost, operatorInfo, executionInfo...). I have record these improvement and once customer require or we have time we will make the result easy to use and more rich.

File: planner/core/common_plans.go [hunk 5]
```diff
@@ -649,6 +649,40 @@ type SelectInto struct {
 	IntoOpt    *ast.SelectIntoOption
 }
 
+// ExplainInfoForEncode store explain info for JSON encode
+type ExplainInfoForEncode struct {
```
Comment (time-and-fate) [inline planner/core/common_plans.go:653]: Why do some fields have `omitempty` but some don't?

File: planner/core/common_plans.go [hunk 6]
```diff
@@ -661,6 +695,7 @@ type Explain struct {
 
 	Rows        [][]string
 	ExplainRows [][]string
+	JSON        JSONSlice
```
Comment (time-and-fate) [inline planner/core/common_plans.go:698]: Not a big issue, but it looks like we don't need to add a new field here, returning the value from `explainFlatPlanInJSONFormat` is enough.

**Comment by dveeden** [comment]
@Icemap maybe this is something to consider for DBeaver after https://github.com/dbeaver/dbeaver/pull/18299 is merged.

This would have made https://dveeden.github.io/ti-explain/ a lot easier to create if it had existed at the time.

File: planner/core/common_plans.go (hunk 5 above)
Comment (fzzf678) [reply planner/core/common_plans.go:653]: > Why do some fields have `omitempty` but some don't?

Some fields in `explain` and some in `explain analyze` and some in both, some common fields don't have `omitempty`. And the `TotalMemoryConsumed` only appears when set `tidb_memory_debug_mode`.

File: planner/core/common_plans.go (hunk 6 above)
Comment (fzzf678) [reply planner/core/common_plans.go]: fixed

File: planner/core/common_plans.go (hunk 5 above)
Comment (time-and-fate) [reply planner/core/common_plans.go:653]: I mean, even for `explain`, we can choose not to print those empty fields, like empty `AccessObject`, `OperatorInfo` and `SubOperators`.

Comment (fzzf678) [reply planner/core/common_plans.go:653]: got it,

**Comment by Icemap** [comment]
> @Icemap maybe this is something to consider for DBeaver after [dbeaver/dbeaver#18299](https://github.com/dbeaver/dbeaver/pull/18299) is merged.
> 
> This would have made https://dveeden.github.io/ti-explain/ a lot easier to create if it had existed at the time.

@dveeden Yes, indeed. If this PR merged. We can use JSON to provide a more useful Execution plan page in DBeaver. For now, the Execution plan page like this:

image

**Review by time-and-fate — APPROVED** [review]
Please also remove the empty fields from the test result.

**Review by chrysan — APPROVED** [review]

**Review by chrysan — APPROVED** [review]

**Pull request merged**